src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs

```diff
@@ -198,27 +221,175 @@ public ImmutableArray<NamespaceOrTypeAndUsingDirective> GetUsingNamespacesOrType
                     throw ExceptionUtilities.UnexpectedValue(declarationSyntax);
             }
 
-            foreach (var declaration in _mergedDeclaration.Declarations)
+            return GetAliasesAndUsings(declarationSyntax).GetUsingNamespacesOrTypes(this, declarationSyntax, basesBeingResolved);
+        }
+
+        private Imports GetGlobalUsingImports(ConsList<TypeSymbol>? basesBeingResolved)
+        {
+            return GetMergedGlobalAliasesAndUsings(basesBeingResolved).Imports;
+        }
+
+        private ImmutableDictionary<string, AliasAndUsingDirective> GetGlobalUsingAliasesMap(ConsList<TypeSymbol>? basesBeingResolved)
+        {
+            return GetMergedGlobalAliasesAndUsings(basesBeingResolved).UsingAliasesMap!;
+        }
+
+        private ImmutableArray<NamespaceOrTypeAndUsingDirective> GetGlobalUsingNamespacesOrTypes(ConsList<TypeSymbol>? basesBeingResolved)
+        {
+            return GetMergedGlobalAliasesAndUsings(basesBeingResolved).UsingNamespacesOrTypes;
+        }
+
+        private MergedGlobalAliasesAndUsings GetMergedGlobalAliasesAndUsings(ConsList<TypeSymbol>? basesBeingResolved, CancellationToken cancellationToken = default)
+        {
+            if (_lazyMergedGlobalAliasesAndUsings is null)
             {
-                var declarationSyntaxRef = declaration.SyntaxReference;
-                if (declarationSyntaxRef.SyntaxTree != declarationSyntax.SyntaxTree)
+                if (!this.IsGlobalNamespace)
                 {
-                    continue;
+                    _lazyMergedGlobalAliasesAndUsings = MergedGlobalAliasesAndUsings.Empty;
                 }
-
-                if (declarationSyntaxRef.GetSyntax() == declarationSyntax)
+                else
                 {
-                    return _aliasesAndUsings[declaration].GetUsingNamespacesOrTypes(this, declarationSyntax, basesBeingResolved);
+                    ImmutableDictionary<string, AliasAndUsingDirective>? mergedAliases = null;
```

Comment:
nit: Why use an immutable structure to accumulate aliases, but a builder for namespaces/types?

Reply:
>  Why use an immutable structure to accumulate aliases, but a builder for namespaces/types?

They are used to build different structures. One is building an ImmutableDictionary, the other an ImmutableArray.

---
In reply to: [595733569](https://github.com/dotnet/roslyn/pull/51855#discussion_r595733569) [](ancestors = 595733569)